### PR TITLE
Solvable synthesis

### PR DESCRIPTION
**Pull request by daneshvar-amrollahi**
Added a new action for synthesizing solvable loops by replacing the defective variables with a fresh variable (if possible). The returned output is a list of solvable programs. 

How to Test:

`python polar.py benchmarks/defective/non-lin-markov-1.prob --synth_solv 2`
`python polar.py benchmarks/defective/squares.prob --synth_solv 1`

The parameter passed to `--synth_solv` is the degree of the invariant (combination) used for the search. 

You may add also add the simultaneous assignment, `n = n + 1, m + 2`  to squares.prob.

### CODE REVIEW

File: cli/argument_parser.py [hunk 1]
```diff
@@ -275,6 +275,13 @@ def __init__(self):
             help="""Get sensitivity of goal with respect to the
             symbolic constants by differentiating the closed form solution"""
         )
+        self.argument_parser.add_argument(
```
Comment (mmsbrggr) [inline cli/argument_parser.py:278]: use mc_comb_deg

Comment (mmsbrggr) [reply cli/argument_parser.py:278]: --synth_solv [optional:list of vars]

File: program/solvable_synthesizer/solvable_synthesizer.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
+from utils import get_unique_var, solve_rec_by_summing, get_terms_with_vars, get_monoms
```
Comment (mmsbrggr) [inline program/solvable_synthesizer/solvable_synthesizer.py:1]: please use mc_comb_finder -> avoid code duplication

File: cli/actions/synthesize_solvable_action.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        print()
+
+        solvable_programs = []
+        combinations = SolvableSynthesizer.synthesize(combination_vars, combination_deg, program)
+        if combinations is None:
+            print(f"No combination found with degree {combination_deg}. Try using higher degrees.")
+            return program
+        else:
+            for combination in combinations:
+                solvable_loop_body = []
```
Comment (mmsbrggr) [inline cli/actions/synthesize_solvable_action.py:43]: Could you move the construction of the program to the SolvableSynthesizer? The CLI actions should do as minimal logic as possible. Doing so we can use the synthesizer in other contexts in the future (without CLI)

**Review by mmsbrggr — CHANGES_REQUESTED** [review]
Functionality is fine. Could you please incorporate the changes mentioned in the comments?

**Comment by daneshvar-amrollahi** [comment]
Fixed bugs with effective variables. 

Input 1:
```
while true:
    x = x + 1
    y = y + x
    z = z**2
end
```
Output 1:
```
while true:
    _t4 = 1 + x
    _t5 = 1 + x + y
    x = _t4
    y = _t5
end

Sample 2:
```
while true:
    s = Bernoulli(1/2)
    if s == 0:
        x, y = x + x*y, (1/3)*x + (2/3)*y + (x*y)
    else:
        x, y = x + y + (2/3)*x*y, 2*y + (2/3)*(x*y)
    end
    z = z**2
    m = m + 1
    n = m + 2
end
```
Output 2:
```
while true:
    _t16 = 1/2
    _s15 = (13/18)*_s15
    _t17 = 1 + m
    _t18 = 3 + m
    s = _t16
    m = _t17
    n = _t18
end
```

File: cli/actions/synthesize_solvable_action.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(self, cli_args: Namespace):
+        self.cli_args = cli_args
+
+    def __call__(self, *args, **kwargs):
+        benchmark = args[0]
+        combination_deg = self.cli_args.mc_comb_deg
+        program = parse_program(benchmark, self.cli_args.transform_categoricals)
+        program = prepare_program(program, self.cli_args)
+
+        if len(program.non_mc_variables) == 0:
```
Comment (mmsbrggr) [inline cli/actions/synthesize_solvable_action.py:21]: Should also work with solvable loops

**Pull request merged**